src/ring_theory/localization.lean

```diff
@@ -1232,36 +1232,55 @@ begin
   rw map_comap,
 end
 
-variables (I) (f : P →+* R)
+variables (I)
 
 /-- For a ring hom `f : P →+* R` and a prime ideal `I` in `R`, the induced ring hom from the
 localization of `P` at `ideal.comap f I` to the localization of `R` at `I` -/
-noncomputable def local_ring_hom : at_prime (ideal.comap f I) →+* at_prime I :=
-(localization.of _).map (λ y, show f y ∈ I.prime_compl, from y.2) (localization.of _)
-
-lemma local_ring_hom_to_map (x : P) :
-  local_ring_hom I f ((localization.of _).to_map x) = (localization.of _).to_map (f x) :=
+noncomputable def local_ring_hom (J : ideal S) [hJ : J.is_prime] (f : R →+* S)
+  (hIJ : ∀ r : R, r ∈ I ↔ f r ∈ J) : at_prime I →+* at_prime J :=
```

Comment:
I think I would phrase `hIJ` in terms of `map` or `comap`. And there can be a lemma that unpacks it into the current form. But now it's hard to apply any lemmas about `map` or `comap`, so you lose access to that api.

Reply:
Do you mean like `hIJ : ∀ r : R, r ∈ I ↔ r ∈ comap f J`? Or rather `hIJ : I = comap f J`?
I think the latter would be a little more difficult to use, e.g. I don't know how to extract the implications `r ∈ I → r ∈ J` from that easily (`ideal.ext_iff` doesn't exist it seems...)

Reply:
`ideal.ext_iff` is just `submodule.ext_iff`. I agree, I think `hIJ : I = comap f J` would be better.

Reply:
On the other hand - we could just leave `local_ring_hom` as it was, and for when you want to use a `J` that isn't defeq to `I` we introduce a
```lean
/-- two equal submonoids have isomorphic localizations
def localization.of_eq {P Q : submonoid M} (h : P = Q): localization P ≃+* localization Q := sorry
```
which you are free to compose with it.

Reply:
I can't find `submodule.ext_iff` either I'm afraid.

Reply:
@eric-wieser, I think introducing `localization.of_eq` or something like that would make the statements of `local_ring_hom_id` and `local_ring_hom_comp` needlessly complicated. Being able to state these cleanly was the original motivation for the change, as they are needed to prove the functor laws of Spec.

Reply:
We can definitely change it to `∀ r : R, r ∈ I ↔ r ∈ comap f J`. I'm not currently convinced that `I = comap f J` would be better.

Reply:
My gut instinct is to use `I = comap f J`. I think it allows for shorter proofs on average, because there is quite a nice API for `map`/`comap`. And when needed, `ext r` will turn it into `r ∈ I ↔ r ∈ comap f J`. Going the other direction feels "annoying" to me.
But as I said, this is just my intuition.

Reply:
Okay, then let's trust that intuition :) 98e54a60b58d722dd08489337faeebc19ae9446d
Exactly, the direction from `I = comap f J` to `r ∈ I ↔ r ∈ comap f J` is a bit annoying. The thing is: There are three places in the code now where I need to go from the hypothesis `I = comap f J` to an implication like `r ∈ I → r ∈ comap f J`, namely to supply the proofs in the arguments of `localization_map.map` and `localization_map.mk'` (see [L1248](https://github.com/leanprover-community/mathlib/blob/98e54a60b58d722dd08489337faeebc19ae9446d/src/ring_theory/localization.lean#L1248), [L1259](https://github.com/leanprover-community/mathlib/blob/98e54a60b58d722dd08489337faeebc19ae9446d/src/ring_theory/localization.lean#L1259) and [L1270](https://github.com/leanprover-community/mathlib/blob/98e54a60b58d722dd08489337faeebc19ae9446d/src/ring_theory/localization.lean#L1270)). I just inserted `set_like.ext_iff.mp` at these places, which feels a bit annoying. Maybe there is a better way to do it. But in the end, it probably doesn't make that much of a difference.

Reply:
I think the reason this is awkward is because `localization_map.map` takes its `hy` argument written elementwise, `∀ y : M, g y ∈ T` instead of `g.range.to_submonoid ≤ T`. Changing that is out of scope for this PR though, and I don't know if it makes things better or worse.